File: app/static/friends.js
// Friends page JavaScript
// Listen for new friend requests using the socket from base.js
document.addEventListener('DOMContentLoaded', function() {
    // Load friends data when page loads
    loadFriendsAndSentData();
    loadFriendRequestsData();
    
    // Wait for socket to be available from base.js
    if (typeof socket !== 'undefined') {
        socket.on('new_notification', function(notification) {
            if (notification.type === 'friend_request') {
                loadFriendRequestsData(); // Reload just the friend requests
            }
        });
    }
});



// Dashboard JavaScript functionality
function openFriendSearch() {
    document.getElementById('friendSearchModal').style.display = 'block';
    document.getElementById('userSearchInput').focus();
}

function closeFriendSearch() {
    document.getElementById('friendSearchModal').style.display = 'none';
    document.getElementById('userSearchInput').value = '';
    document.getElementById('searchResults').innerHTML = '';
}

let searchTimeout;

function searchUsers() {
    clearTimeout(searchTimeout);
    const query = document.getElementById('userSearchInput').value.trim();
    
    if (query.length < 2) {
        document.getElementById('searchResults').innerHTML = '';
        return;
    }
    
    searchTimeout = setTimeout(() => {
        fetch(`/search_users?q=${encodeURIComponent(query)}`)
            .then(response => response.json())
            .then(users => {
                displaySearchResults(users);
            })
            .catch(error => {
                console.error('Error searching users:', error);
            });
    }, 300);
}

function displaySearchResults(users) {
    const resultsContainer = document.getElementById('searchResults');
    
    if (users.length === 0) {
        resultsContainer.innerHTML = '<div class="search-result-item no-results">No users found</div>';
        return;
    }
    
    resultsContainer.innerHTML = users.map(user => {
        let buttonHtml = '';
        let statusText = '';
        
        switch(user.status) {
            case 'none':
                buttonHtml = `<button class="btn btn-secondary btn-sm" onclick="sendFriendRequest(${user.id})">Add Friend</button>`;
                break;
            case 'friends':
                statusText = '<span class="status-badge friends">Friends</span>';
                break;
            case 'request_sent':
                statusText = '<span class="status-badge pending">Request Sent</span>';
                break;
            case 'request_received':
                statusText = '<span class="status-badge received">Request Received</span>';
                break;
        }
        
        return `
            <div class="search-result-item">
                <div class="user-info">
                    <strong>${user.username}</strong>
                    ${statusText}
                </div>
                <div class="user-actions">
                    ${buttonHtml}
                </div>
            </div>
        `;
    }).join('');
}

function sendFriendRequest(userId) {
    fetch('/send_friend_request', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({ user_id: userId })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            alert(data.message);
            searchUsers(); // Refresh search results
        } else {
            alert('Error: ' + data.message);
        }
    })
    .catch(error => {
        console.error('Error sending friend request:', error);
        alert('Error sending friend request');
    });
}

// Friends functionality removed from dashboard

function loadFriendsAndSentData() {
    const combinedList = document.getElementById('friends-combined-list');
    const totalCount = document.getElementById('friends-total-count');
    
    // Load both friends and sent requests simultaneously
    Promise.all([
        fetch('/api/friends').then(response => response.json()),
        fetch('/api/sent_requests').then(response => response.json())
    ])
    .then(([friends, sentRequests]) => {
        const totalItems = friends.length + sentRequests.length;
        totalCount.textContent = totalItems;
        
        if (totalItems === 0) {
            combinedList.innerHTML = '<div class="no-data">You don\'t have any friends yet. Use the "Add Friend" button to send friend requests!</div>';
            return;
        }
        
        let html = '';
        
        // Add accepted friends first
        friends.forEach(friend => {
            html += `
                <div class="friend-item">
                    <div class="friend-info">
                        <a href="/profile/${friend.username}" class="friend-name-link">
                            <strong>${friend.username}</strong>
                        </a>
                        <span class="status-badge ${friend.is_online ? 'online' : 'offline'}">
                            ${friend.is_online ? 'Online' : 'Offline'}
                        </span>
                    </div>
                    <div class="friend-actions">
                        <button class="btn btn-remove btn-sm" onclick="removeFriend(${friend.id}, '${friend.username}', this)">Remove</button>
                    </div>
                </div>
            `;
        });
        
        // Add sent requests with pending status
        sentRequests.forEach(request => {
            html += `
                <div class="friend-item sent-request-item" data-friendship-id="${request.friendship_id}">
                    <div class="friend-info">
                        <strong>${request.username}</strong>
                        <span class="status-badge pending">Pending</span>
                    </div>
                    <div class="request-actions">
                        <button class="btn btn-cancel btn-sm" onclick="cancelFriendRequest(${request.friendship_id}, this)">Cancel</button>
                    </div>
                </div>
            `;
        });
        
        combinedList.innerHTML = html;
    })
    .catch(error => {
        console.error('Error loading friends and sent requests:', error);
        combinedList.innerHTML = '<div class="error">Error loading data</div>';
    });
}

function loadFriendRequestsData() {
    const requestsList = document.getElementById('friend-requests-list');
    const requestsCount = document.getElementById('requests-count');
    
    fetch('/api/friend_requests')
        .then(response => response.json())
        .then(requests => {
            requestsCount.textContent = requests.length;
            
            if (requests.length === 0) {
                requestsList.innerHTML = '<div class="no-data">No pending friend requests.</div>';
                return;
            }
            
            requestsList.innerHTML = requests.map(request => `
                <div class="friend-request-item" data-friendship-id="${request.friendship_id}">
                    <div class="friend-info">
                        <strong>${request.username}</strong>
                        <span class="request-text">sent you a friend request</span>
                    </div>
                    <div class="request-actions">
                        <button class="btn btn-accept btn-sm" onclick="respondToFriendRequest(${request.friendship_id}, 'accept', this)">Accept</button>
                        <button class="btn btn-reject btn-sm" onclick="respondToFriendRequest(${request.friendship_id}, 'reject', this)">Reject</button>
                    </div>
                </div>
            `).join('');
        })
        .catch(error => {
            console.error('Error loading friend requests:', error);
            requestsList.innerHTML = '<div class="error">Error loading friend requests</div>';
        });
}

function respondToFriendRequest(friendshipId, response, buttonElement) {
    const requestItem = buttonElement.closest('.friend-request-item');
    const buttons = requestItem.querySelectorAll('button');
    
    // Disable buttons during request
    buttons.forEach(btn => btn.disabled = true);
    
    fetch('/respond_friend_request', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            friendship_id: friendshipId,
            response: response
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            if (response === 'accept') {
                // Show success message and refresh both lists
                showToast('Friend request accepted!');
                loadFriendsAndSentData(); // Refresh combined friends list
                loadFriendRequestsData(); // Refresh requests list
            } else {
                // Remove the request item
                requestItem.remove();
                // Update count
                const requestsCount = document.getElementById('requests-count');
                requestsCount.textContent = parseInt(requestsCount.textContent) - 1;
                showToast('Friend request rejected');
            }
        } else {
            // Re-enable buttons if there was an error
            buttons.forEach(btn => btn.disabled = false);
            showToast('Error: ' + data.message);
        }
    })
    .catch(error => {
        console.error('Error responding to friend request:', error);
        buttons.forEach(btn => btn.disabled = false);
        showToast('Error processing request');
    });
}

function cancelFriendRequest(friendshipId, buttonElement) {
    const requestItem = buttonElement.closest('.sent-request-item');
    const button = requestItem.querySelector('button');
    
    // Disable button during request
    button.disabled = true;
    button.textContent = 'Cancelling...';
    
    fetch('/cancel_friend_request', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            friendship_id: friendshipId
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            // Refresh the entire combined list to update counts
            loadFriendsAndSentData();
            showToast('Friend request cancelled');
        } else {
            // Re-enable button if there was an error
            button.disabled = false;
            button.textContent = 'Cancel';
            showToast('Error: ' + data.message);
        }
    })
    .catch(error => {
        console.error('Error cancelling friend request:', error);
        button.disabled = false;
        button.textContent = 'Cancel';
        showToast('Error cancelling request');
    });
}

function removeFriend(friendUserId, friendUsername, buttonElement) {
    // Confirm before removing friend
    if (!confirm(`Are you sure you want to remove ${friendUsername} from your friends?`)) {
        return;
    }
    
    const friendItem = buttonElement.closest('.friend-item');
    const button = friendItem.querySelector('.btn-remove');
    
    // Disable button during request
    button.disabled = true;
    button.textContent = 'Removing...';
    
    fetch('/remove_friend', {
        method: 'POST',
        headers: {
            'Content-Type': 'application/json',
        },
        body: JSON.stringify({
            friend_user_id: friendUserId
        })
    })
    .then(response => response.json())
    .then(data => {
        if (data.success) {
            // Refresh the entire combined list to update counts
            loadFriendsAndSentData();
            showToast(`${friendUsername} removed from friends`);
        } else {
            // Re-enable button if there was an error
            button.disabled = false;
            button.textContent = 'Remove';
            showToast('Error: ' + data.message);
        }
    })
    .catch(error => {
        console.error('Error removing friend:', error);
        button.disabled = false;
        button.textContent = 'Remove';
        showToast('Error removing friend');
    });
}

// Event handlers
window.onclick = function(event) {
    const friendSearchModal = document.getElementById('friendSearchModal');
    
    if (event.target === friendSearchModal) {
        closeFriendSearch();
    }
}

File: app/routes/friends.py

from flask import Blueprint, render_template, request, session, jsonify
from ..helpers import create_notification, clean_notification_data
from ..models import User, Notification, Messenger
from ..decorators import login_required
from .. import db
from ..services import user_service, friendship_service, notification_service

bp_friends = Blueprint("bp_friends", __name__)

@bp_friends.route('/friends')
@login_required
def friends():
    current_user_id = session['user_id']
    
    friends = user_service.get_user_friends(current_user_id)

    received_requests = friendship_service.get_received_friend_requests(current_user_id).all()
    
    return render_template('friends.html', friends=friends, received_requests=received_requests)

@bp_friends.route('/search_users')
@login_required
def search_users():
    query = request.args.get('q', '')
    if len(query) < 2:
        return jsonify([])
    
    users = User.query.filter(
        User.username.contains(query),
        User.id != session['user_id']
    ).limit(10).all()
    
    current_user_id = session['user_id']
    results = []
    
    for user in users:
        status = friendship_service.get_friendship_status(current_user_id, user.id)
        results.append({
            'id': user.id,
            'username': user.username,
            'status': status
        })
    
    return jsonify(results)

@bp_friends.route('/send_friend_request', methods=['POST'])
@login_required
def send_friend_request():
    requested_user_id = request.json.get('user_id')
    current_user_id = session['user_id']
    
    if requested_user_id == current_user_id:
        return jsonify({'success': False, 'message': 'Cannot send friend request to yourself'})
    
    try:
        new_friendship = friendship_service.send_friend_request(current_user_id, requested_user_id)
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})
    
    # Create notification for the requested user
    requester = User.query.get(current_user_id)
    create_notification(
        requested_user_id,
        'friend_request',
        f'{requester.username} sent you a friend request',
        {'friendship_id': new_friendship.id, 'requester_username': requester.username}
    )
    
    return jsonify({'success': True, 'message': 'Friend request sent!'})

@bp_friends.route('/respond_friend_request', methods=['POST'])
@login_required
def respond_friend_request():
    friendship_id = request.json.get('friendship_id')
    response = request.json.get('response')
    current_user_id = session['user_id']
    
    try:
        if response == 'accept':
            friendship = friendship_service.accept_friend_request(friendship_id, current_user_id)

            new_messenger = Messenger(
                first_user_id=friendship.requested_id,
                second_user_id=friendship.requester_id,
            )
            db.session.add(new_messenger)
            db.session.commit()

            # Notify the requester that their request was accepted
            accepter = User.query.get(current_user_id)
            create_notification(
                friendship.requester_id,
                'friend_request_accepted',
                f'{accepter.username} accepted your friend request',
                {'friendship_id': friendship.id, 'accepter_username': accepter.username}
            )

            return jsonify({'success': True, 'message': 'Friend request accepted!'})

        elif response == 'reject':

            try:
                friend_id = friendship_service.reject_friend_request(friendship_id, current_user_id)

                notification_service.remove_friendship_history(
                    session["user_id"],
                    friend_id,
                    friendship_id
                )

            except Exception as e:
                return jsonify({'success': False, 'message': str(e)})

            return jsonify({'success': True, 'message': 'Friend request rejected'})

    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})

@bp_friends.route('/cancel_friend_request', methods=['POST'])
@login_required
def cancel_friend_request():
    friendship_id = request.json.get('friendship_id')
    current_user_id = session['user_id']
    
    try:
        friend_id = friendship_service.cancel_friend_request(friendship_id, current_user_id)

        notification_service.remove_friendship_history(
            session["user_id"],
            friend_id,
            friendship_id
        )

    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})
    
    return jsonify({'success': True, 'message': 'Friend request cancelled'})

@bp_friends.route('/remove_friend', methods=['POST'])
@login_required
def remove_friend():
    friend_user_id = request.json.get('friend_user_id')
    current_user_id = session['user_id']
    
    if friend_user_id == current_user_id:
        return jsonify({'success': False, 'message': 'Cannot remove yourself'})
    
    try:
        friendship_service.remove_friend(current_user_id, friend_user_id)
    except Exception as e:
        return jsonify({'success': False, 'message': str(e)})
    
    return jsonify({'success': True, 'message': 'Friend removed successfully'})

# ============================================================================
# API ROUTES (JSON DATA)
# ============================================================================

@bp_friends.route('/api/friends')
@login_required
def get_friends_data():
    current_user_id = session['user_id']
    
    # Get accepted friends
    friends = friendship_service.get_friends_query(current_user_id).all()
    
    friends_data = []
    for user, friendship in friends:
        friends_data.append({
            'id': user.id,
            'username': user.username,
            'is_online': user.is_online,
            'last_seen': user.last_seen.isoformat() if user.last_seen else None
        })
    
    return jsonify(friends_data)

@bp_friends.route('/api/friend_requests')
@login_required
def get_friend_requests_data():
    current_user_id = session['user_id']
    
    # Get received friend requests
    received_requests = friendship_service.get_received_friend_requests(current_user_id).all()
    
    requests_data = []
    for user, friendship in received_requests:
        requests_data.append({
            'id': user.id,
            'username': user.username,
            'friendship_id': friendship.id,
            'created_at': friendship.created_at.isoformat()
        })
    
    return jsonify(requests_data)

@bp_friends.route('/api/sent_requests')
@login_required
def get_sent_requests_data():
    current_user_id = session['user_id']
    
    # Get sent friend requests that are still pending
    sent_requests = friendship_service.get_sent_friend_requests(current_user_id).all()
    
    requests_data = []
    for user, friendship in sent_requests:
        requests_data.append({
            'id': user.id,
            'username': user.username,
            'friendship_id': friendship.id,
            'created_at': friendship.created_at.isoformat()
        })
    
    return jsonify(requests_data)